dbms/src/Storages/KVStore/MultiRaft/RegionSerde.cpp

```diff
@@ -0,0 +1,279 @@
+                        actual_extension_count,
+                        buf,
+                        magic_enum::enum_underlying(RegionPersistExtension::ReservedForTest),
+                        s.data(),
+                        s.size());
+                }
+                if (value & 2)
+                {
+                    std::string s = "kkk";
+                    total_size += writePersistExtension(
+                        actual_extension_count,
+                        buf,
+                        UNUSED_EXTENSION_NUMBER_FOR_TEST,
+                        s.data(),
+                        s.size());
+                }
+                if (value & 4)
+                {
+                    std::string s = "zzz";
+                    total_size += writePersistExtension(
+                        actual_extension_count,
+                        buf,
+                        UNUSED_EXTENSION_NUMBER_FOR_TEST,
+                        s.data(),
+                        s.size());
+                }
+            }
+        });
+        RUNTIME_CHECK(
+            expected_extension_count == actual_extension_count,
+            expected_extension_count,
+            actual_extension_count);
+
+        // serialize data
+        total_size += data.serialize(buf);
+    }
+
+    return {total_size, applied_index};
+}
+
+/// Currently supports:
+/// 1. Vx -> Vy where x >= 2, y >= 3
+/// 2. Vx -> V2 where x >= 2, in 7.5.0
+/// 3. Vx -> V2 where x >= 2, in later 7.5
+RegionPtr Region::deserialize(ReadBuffer & buf, const TiFlashRaftProxyHelper * proxy_helper)
+{
+    const auto binary_version = readBinary2<UInt32>(buf);
+    auto current_version = Region::CURRENT_VERSION;
+    fiu_do_on(FailPoints::force_region_read_version, {
+        if (auto v = FailPointHelper::getFailPointVal(FailPoints::force_region_read_version); v)
+        {
+            current_version = std::any_cast<UInt64>(v.value());
+            LOG_WARNING(
+                Logger::get(),
+                "Failpoint force_region_read_version set region binary version, value={}",
+                current_version);
+        }
+    });
+
+    if (current_version <= 1 && binary_version > current_version)
+    {
+        // Conform to https://github.com/pingcap/tiflash/blob/43f809fffde22d0af4c519be4546a5bf4dde30a2/dbms/src/Storages/KVStore/Region.cpp#L197
+        // When downgrade from x(where x > 1) -> 1, the old version will throw with "unexpected version".
+        // So we will also throw here.
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Don't support downgrading from {} to {}",
+            binary_version,
+            current_version);
+    }
+    const auto binary_version_decoded = magic_enum::enum_cast<RegionPersistVersion>(binary_version);
+    if (!binary_version_decoded.has_value())
+    {
+        LOG_DEBUG(DB::Logger::get(), "Maybe downgrade from {} to {}", binary_version, current_version);
```

Comment:
If there is 30000 region, we could print 30000 times. However, I think this information may be not necessary. Because downgrading may be a common operation later.